test/integration/controller/jobset_controller_test.go

```diff
@@ -666,6 +669,291 @@ var _ = ginkgo.Describe("JobSet controller", func() {
+							{Name: "replicated-job-b",
+								Ready:     0,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    0,
+								Suspended: 3,
+							},
+							{
+								Name:      "replicated-job-a",
+								Ready:     0,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    0,
+								Suspended: 1,
+							},
+						})
+					},
+				},
+				{
+					jobUpdateFn: func(jobList *batchv1.JobList) {
+						readyJob(&jobList.Items[0])
+					},
+				},
+				{
+					checkJobSetState: func(js *jobset.JobSet) {
+						matchJobSetReplicatedStatus(js, []jobset.ReplicatedJobStatus{
+							{Name: "replicated-job-b",
+								Ready:     0,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    0,
+								Suspended: 0,
+							},
+							{
+								Name:      "replicated-job-a",
+								Ready:     1,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    0,
+								Suspended: 0,
+							},
+						})
+					},
+				},
+				{
+					checkJobSetCondition: testutil.JobSetStartupPolicyComplete,
+				},
+			},
+		}),
+		ginkgo.Entry("startupPolicy; replicated-job-a not ready than replicated-job-b should not run", &testCase{
+			makeJobSet: func(ns *corev1.Namespace) *testing.JobSetWrapper {
+				return testJobSet(ns).
+					StartupPolicy(&jobset.StartupPolicy{
+						StartupPolicyOrder: jobset.InOrder,
+					})
+			},
+			skipCreationCheck: true,
+			updates: []*update{
+				{
+					checkJobSetState: func(js *jobset.JobSet) {
+						matchJobSetReplicatedStatus(js, []jobset.ReplicatedJobStatus{
+							{Name: "replicated-job-b",
+								Ready:     0,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    0,
+								Suspended: 0,
+							},
+							{
+								Name:      "replicated-job-a",
+								Ready:     0,
+								Succeeded: 0,
+								Failed:    0,
+								Active:    1,
+								Suspended: 0,
+							},
+						})
+					},
+					checkJobSetCondition: testutil.JobSetStartupPolicyNotFinished,
+				},
+			},
+		}),
+		ginkgo.Entry("startupPolicy; success policy restart; replicated-job-a ready than replicated-job-b should run", &testCase{
+			makeJobSet: func(ns *corev1.Namespace) *testing.JobSetWrapper {
+				return testJobSet(ns).
+					FailurePolicy(&jobset.FailurePolicy{MaxRestarts: 1}).
+					StartupPolicy(&jobset.StartupPolicy{
+						StartupPolicyOrder: jobset.InOrder,
+					})
+			},
+			skipCreationCheck: true,
+			updates: []*update{
+				{
+					jobUpdateFn: func(jobList *batchv1.JobList) {
+						readyJob(&jobList.Items[0])
+						readyJob(&jobList.Items[1])
+						readyJob(&jobList.Items[2])
+						readyJob(&jobList.Items[3])
+					},
+				},
+				{
+					jobUpdateFn: func(jobList *batchv1.JobList) {
+						completeJob(&jobList.Items[0])
+						failJob(&jobList.Items[1])
+					},
+					checkJobSetState: func(js *jobset.JobSet) {
+						ginkgo.By("checking all jobs are recreated")
+						gomega.Eventually(checkJobsRecreated, timeout, interval).WithArguments(js, 1).Should(gomega.Equal(true))
```

Comment:
Not sure if this is relevant anymore. I couldn't use this one directly because startup policy and recration interfer with one another.

ie if startup poliy is specified we recreate in order so this can be run after a startup policy success.